packages/remix-react/components.tsx

```diff
@@ -666,52 +666,108 @@ export function Meta() {
         typeof routeModule.meta === "function"
           ? routeModule.meta({ data, parentsData, params, location })
           : routeModule.meta;
-      Object.assign(meta, routeMeta);
+      if (routeMeta) {
+        meta = processMeta(meta, routeMeta);
+      }
     }
 
     parentsData[routeId] = data;
   }
 
   return (
     <>
-      {Object.entries(meta).map(([name, value]) => {
-        if (!value) {
+      {meta.map((metaDescriptor) => {
+        if (!metaDescriptor) {
           return null;
         }
 
-        if (["charset", "charSet"].includes(name)) {
-          return <meta key="charset" charSet={value as string} />;
+        if (metaDescriptor.hasOwnProperty("charset")) {
+          return <meta key="charSet" charSet={metaDescriptor.charset} />;
         }
-
-        if (name === "title") {
-          return <title key="title">{value}</title>;
+        if (metaDescriptor.hasOwnProperty("charSet")) {
+          return <meta key="charSet" charSet={metaDescriptor.charSet} />;
         }
-
-        // Open Graph tags use the `property` attribute, while other meta tags
-        // use `name`. See https://ogp.me/
-        let isOpenGraphTag = name.startsWith("og:");
-        return [value].flat().map((content) => {
-          if (isOpenGraphTag) {
-            return (
-              <meta
-                content={content as string}
-                key={name + content}
-                property={name}
-              />
-            );
-          }
-
-          if (typeof content === "string") {
-            return <meta content={content} name={name} key={name + content} />;
-          }
-
-          return <meta key={name + JSON.stringify(content)} {...content} />;
-        });
+        if (metaDescriptor.hasOwnProperty("title")) {
+          return <title key="title">{metaDescriptor.title}</title>;
+        }
+        return (
+          <meta {...metaDescriptor} key={JSON.stringify(metaDescriptor)} />
+        );
       })}
     </>
   );
 }
 
+/*
+ * This function processes a meta descriptor and adds it to the meta array.
+ * This converts the original object syntax to new array syntax.
+ */
+export function processMeta(
```

Comment:
change `HtmlMetaDescriptor` to this:

```ts
export type HtmlMetaDescriptor = {
  key?: string;
  name?: string;
  property?: string;
  title?: string;
  content?: string;
} & Record<string, string | string[] | null | undefined>;
```

and try this for `processMeta`. it sets `key` here

```tsx
export function processMeta(
  meta: MetaMap,
  routeMeta: HtmlMetaDescriptor | HtmlMetaDescriptor[]
) {
  let items: HtmlMetaDescriptor[] = Array.isArray(routeMeta)
    ? routeMeta
    : Object.entries(routeMeta)
        .map(([key, value]) => {
          if (!value) return [];

          let propertyName = key.startsWith("og:") ? "property" : "name";
          return key === "title"
            ? { key: "title", content: assertString(value) }
            : ["charset", "charSet"].includes(key)
            ? { key: "charset", content: assertString(value) }
            : Array.isArray(value)
            ? value.map((content) => ({
                key: `${key}.${content}`,
                [propertyName]: key,
                content,
              }))
            : { key, [propertyName]: key, content: value };
        })
        .flat();

  items.forEach((item) => {
    let [key, value] = computeKey(item);
    // only set if key doesn't exist (i.e. let the furthest route win)
    if (!meta.has(key)) {
      meta.set(key, value);
    }
  });
}
```

other needed helpers:

```tsx
function assertString(value: string | string[]): string {
  if (typeof value !== "string") {
    throw new Error("Expected string, got an array of strings");
  }
  return value;
}

function generateKey(item: HtmlMetaDescriptor) {
  console.warn("No key provided to meta", item);
  return JSON.stringify(item); // generate a reasonable key
}

function computeKey(item: HtmlMetaDescriptor): [string, HtmlMetaDescriptor] {
  let {
    name,
    property,
    title,
    key = name ?? property ?? title ?? generateKey(item),
    ...rest
  } = item;

  return [key, { name, property, title, ...rest }];
}
```

Reply:
I've made these changes, but a couple of comments:

1) I was trying not to change `HtmlMetaDescriptor` since it's a public API. Your version no longer supports the extended object syntax (ie., httpEquiv). My understanding is that's going to be reverted anyway.

2) The array syntax no longer supports the shorthand for `{ title: "my title"}` instead requiring `{ key: "title", content"my title" }`. Same with charset. Probably need Ryan to weigh-in on this change.

3) When you're merging meta map, you said only set if key doesn't exist. However, `processMap` is called from the root to the child route, so without subsequent sets, you can't override. I just removed the check.

With these changes, my local tests work. Still having issues with the integration test.